kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceAction.java

```diff
@@ -0,0 +1,363 @@
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.kafkarest.resources.v3;
+
+import static java.util.Objects.requireNonNull;
+
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.protobuf.ByteString;
+import io.confluent.kafka.serializers.subject.strategy.SubjectNameStrategy;
+import io.confluent.kafkarest.controllers.ProduceController;
+import io.confluent.kafkarest.controllers.RecordSerializer;
+import io.confluent.kafkarest.controllers.SchemaManager;
+import io.confluent.kafkarest.entities.EmbeddedFormat;
+import io.confluent.kafkarest.entities.ProduceResult;
+import io.confluent.kafkarest.entities.RegisteredSchema;
+import io.confluent.kafkarest.entities.v3.ProduceRequest;
+import io.confluent.kafkarest.entities.v3.ProduceRequest.ProduceRequestData;
+import io.confluent.kafkarest.entities.v3.ProduceRequest.ProduceRequestHeader;
+import io.confluent.kafkarest.entities.v3.ProduceResponse;
+import io.confluent.kafkarest.entities.v3.ProduceResponse.ProduceResponseData;
+import io.confluent.kafkarest.exceptions.StatusCodeException;
+import io.confluent.kafkarest.response.StreamingResponse;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collector;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response.Status;
+
+@Path("/v3/clusters/{clusterId}/topics/{topicName}/records")
+public final class ProduceAction {
+
+  private static final Collector<
+      ProduceRequestHeader,
+      ImmutableMultimap.Builder<String, Optional<ByteString>>,
+      ImmutableMultimap<String, Optional<ByteString>>> PRODUCE_REQUEST_HEADER_COLLECTOR =
+      Collector.of(
+          ImmutableMultimap::builder,
+          (builder, header) -> builder.put(header.getName(), header.getValue()),
+          (left, right) -> left.putAll(right.build()),
+          ImmutableMultimap.Builder::build);
+
+  private final Provider<SchemaManager> schemaManager;
+  private final Provider<RecordSerializer> recordSerializer;
+  private final Provider<ProduceController> produceController;
```

Comment:
If we could, we would also be able to just inject the instance directly instead of injection the provider :). TL;DR, the request handling pipeline currently looks something like:

URL-matching -> ctor called -> auth -> method called.

We can only call the provider after `auth`. We will fix this when we re-implement `auth` using dependency injection. We will only do that after we rewrite the consume API, just because it will be much easier then.

Reply:
Thanks for unwrapping this for me (no pun intended)!
Can you please add a comment explaining that?

Reply:
Not adding the comment since the is the usual/expected usage of Provider for injection.